src/metadata/trackmetadatataglib.cpp

```diff
@@ -0,0 +1,768 @@
+        QByteArray data(
+                QByteArray::fromBase64(
+                        tag.fieldListMap()["COVERART"].toString().toCString()));
+        coverArt = QImage::fromData(data);
+    }
+    return coverArt;
+}
+
+QImage readMP4TagCover(/*const*/TagLib::MP4::Tag& tag) {
+    QImage coverArt;
+    if (tag.itemListMap().contains("covr")) {
+        TagLib::MP4::CoverArtList coverArtList =
+                tag.itemListMap()["covr"].toCoverArtList();
+        TagLib::ByteVector data = coverArtList.front().data();
+        coverArt = QImage::fromData(
+                reinterpret_cast<const uchar *>(data.data()), data.size());
+    }
+    return coverArt;
+}
+
+namespace {
+
+void replaceID3v2Frame(TagLib::ID3v2::Tag* pTag, TagLib::ID3v2::Frame* pFrame) {
+    pTag->removeFrames(pFrame->frameID());
+    pTag->addFrame(pFrame);
+}
+
+void writeID3v2TextIdentificationFrame(TagLib::ID3v2::Tag* pTag,
+        const TagLib::ByteVector &id, const QString& text) {
+    TagLib::String::Type textType;
+    // For an overview of the character encodings supported by
+    // the different ID3v2 versions please refer to the following
+    // resources:
+    // http://en.wikipedia.org/wiki/ID3#ID3v2
+    // http://id3.org/id3v2.3.0
+    // http://id3.org/id3v2.4.0-structure
+    if (4 <= pTag->header()->majorVersion()) {
+        // For ID3v2.4.0 or higher prefer UTF-8, because it is a
+        // very compact representation for common cases and it is
+        // independent of the byte order.
+        textType = TagLib::String::UTF8;
+    } else {
+        // For ID3v2.3.0/ID3v2.2.0 use UCS-2 (UTF-16 encoded Unicode
+        // with BOM), because UTF-8 and UTF-16BE are only supported
+        // since ID3v2.4.0 and the alternative ISO-8859-1 does not
+        // cover all Unicode characters.
+        textType = TagLib::String::UTF16;
+    }
+    QScopedPointer<TagLib::ID3v2::TextIdentificationFrame> pNewFrame(
+            new TagLib::ID3v2::TextIdentificationFrame(id, textType));
+    pNewFrame->setText(toTagLibString(text));
+    replaceID3v2Frame(pTag, pNewFrame.data());
+    // Now the plain pointer in pNewFrame is owned and
+    // managed by pTag. We need to release the ownership
+    // to avoid double deletion!
+    pNewFrame.take();
+}
+
+} // anonymous namespace
+
+bool writeTag(TagLib::Tag* pTag, const TrackMetadata& trackMetadata) {
+    if (NULL == pTag) {
+        return false;
+    }
+
+    pTag->setArtist(toTagLibString(trackMetadata.getArtist()));
+    pTag->setTitle(toTagLibString(trackMetadata.getTitle()));
+    pTag->setAlbum(toTagLibString(trackMetadata.getAlbum()));
+    pTag->setGenre(toTagLibString(trackMetadata.getGenre()));
+    pTag->setComment(toTagLibString(trackMetadata.getComment()));
+
+    // Only write the year into the common tag if the corresponding
+    // field contains a 4-character numeric string.
+    const QString yearString(trackMetadata.getYear().trimmed());
```

Comment:
Will we loose the YYYY-MM-DD info here, if the track has any? 
Currently we can edit them like that in the property window but they are hidden in the library.